gwcs/wcs.py

```diff
@@ -634,3 +641,223 @@ def fix_inputs(self, fixed):
+    outputs (x, y)
+    """
+    llsqfitter = LinearLSQFitter()
+    # The case of one pass with the specified polynomial degree
+    if degree is None:
+        deglist = range(1,10)
+    else:
+        deglist = [degree]
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
+        fit_poly_x = llsqfitter(poly_x, u, v, x)
+        fit_poly_y = llsqfitter(poly_y, u, v, y)
+        max_resid, rms = _compute_distance_residual(x, y,
+                                                    fit_poly_x(u, v), fit_poly_y(u, v))
+        if verbose:
+            print(f'Degree = {degree}, max_resid = {max_resid}, rms = {rms}')
+        if ((max_error is not None and max_resid < max_error and max_rms is None)
+             or (max_error is not None and max_resid < max_error and max_rms is not None
+                 and rms < max_rms)
+             or (max_error is None and max_rms is not None and rms > max_rms)):
+            if verbose:
+                print('terminating condition met')
+            break
+        if deg == 9:
+            raise RuntimeError("Unable to achieve required fit accuracy with SIP polynomials")
+    return fit_poly_x, fit_poly_y, max_resid, rms
+
+
+def _compute_distance_residual(undist_x, undist_y, fit_poly_x, fit_poly_y):
+    """
+    Compute the distance residuals and return the rms and maximum values.
+    """
+    dist = np.sqrt((undist_x - fit_poly_x)**2 + (undist_y - fit_poly_y)**2)
+    max_resid = dist.max()
+    rms = np.sqrt((dist**2).sum() / dist.size)
+    return max_resid, rms
+
+
+def _reform_poly_coefficients(fit_poly_x, fit_poly_y):
+    """
+    The fit polynomials must be recombined to align with the SIP decomposition
+
+    The result is the f(u,v) and g(u,v) polynomials, and the CD matrix.
+    """
+    # Extract values for CD matrix and recombining
+    c11 = fit_poly_x.c1_0.value
+    c12 = fit_poly_x.c0_1.value
+    c21 = fit_poly_y.c1_0.value
+    c22 = fit_poly_y.c0_1.value
+    sip_poly_x = fit_poly_x.copy()
+    sip_poly_y = fit_poly_y.copy()
+    # Force low order coefficients to be 0 as defined in SIP
+    sip_poly_x.c0_0 = 0
+    sip_poly_y.c0_0 = 0
+    sip_poly_x.c1_0 = 0
+    sip_poly_x.c0_1 = 0
+    sip_poly_y.c1_0 = 0
+    sip_poly_y.c0_1 = 0
+    det = c11 * c22 - c12 * c21
+    ff = c22 / det
+    fg = c12 / det
+    gg = c11 / det
+    gf = c21 / det
+    degree = fit_poly_x.degree
+    # Now loop through all remaining coefficients
+    for i in range(0, degree + 1):
+        for j in range(0, degree + 1):
+            if (i + j > 1) and (i + j < degree + 1):
+                old_x = getattr(fit_poly_x, f'c{i}_{j}').value
+                old_y = getattr(fit_poly_y, f'c{i}_{j}').value
+                new_x = ff * old_x - fg * old_y
+                setattr(sip_poly_x, f'c{i}_{j}', new_x)
+                new_y = gg * old_y - gf * old_x
+                setattr(sip_poly_y, f'c{i}_{j}', new_y)
+    cd = ((c11, c12), (c21, c22))
```

Comment:
Ugh! Is there a reason you are not using np.linalg for the inverse and np.dot for the new coefficients?
The math is corretc but it's going to be more concise and easy to read if using numpy.
Also an error should be raised if the determinant is 0.

Reply:
If the determinant is 0, there is something very wrong with the GWCS. So an error can be raised, but they really should have noticed the problem earlier.

Hey, I'm old fashioned and believe in counting on my fingers. But more seriously, yes, it is simpler in this case, but not for where the inverse transform is computed, mainly because of the index games that must be played to use the standard functions. If it were a 3x3 case, sure, but 2x2 is pretty simple in its explicit form. I changed the one you marked, but left the other as is.

Reply:
I agree that for a `2x2` doing this on fingers might be OK. The only bad thing about it is possible accuracy loss. A better way would be to do inversion using pivot elements or some other more stable technique which, I am sure, is used in `numpy.linalg`. Hopefully in practice we will not encounter ill-conditioned matrices but if you want to be more prepared for weird situations then `numpy.linalg` is the way to go.

Reply:
@mcara I suspect this is a case of letting a rule of thumb become an absolute, even when isn't appropriate. As I understand it pivoting is used to eliminate numerical problems with gaussian elimination, but gaussian elimination isn't used in the 2x2 case and I'm fairly confident that the direct expressions here are as good as anything (I'm happy to be proved wrong). 

Furthermore, remember this is for "Simple Image Polynomial". The magic word is image, and that is presumably in a undistorted frame. The axes are presumably nearly orthogonal and that should never lead to a problem matrix. If there is an ill-conditioned matrix, someone is already seriously screwed in their model (i.e., something is already very wrong).